sw/device/lib/dif/dif_spi_device_test.cc

```diff
@@ -0,0 +1,832 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+extern "C" {
+#include "sw/device/lib/dif/dif_spi_device.h"
+
+#include "spi_device_regs.h"  // Generated.
+}  // extern "C"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+namespace dif {
+namespace {
+using ::mock_mmio::LeInt;
+using ::mock_mmio::MmioTest;
+using ::mock_mmio::MockDevice;
+using ::testing::StrictMock;
+using ::testing::Test;
+
+// Convenience function for assembling a phased FIFO pointer.
+uintptr_t FifoPtr(uintptr_t offset, bool phase) {
+  return offset | (static_cast<uintptr_t>(phase) << 11);
+}
+
+// Convenience function for generating a vector full of noisy data.
+std::vector<char> MakeBlob(size_t len) {
+  std::vector<char> buf;
+  buf.resize(len);
+  uint8_t val = 1;
+  for (auto &c : buf) {
+    c = val;
+    val *= 31;
+  }
+  return buf;
+}
+
+class SpiTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+  static const uint16_t kFifoLen = 0x400;
+
+  dif_spi_device_t spi_ = {
+      /*base_addr=*/dev().region(),
```

Comment:
Didn't we allow designated initializers in C++?

Reply:
I don't think so....?

Reply:
I think this requires some clarification in the [guideline](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) because [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Nonstandard_Extensions) says no.

Reply:
We do not yet allow designated initializers, as they're not part of C++14 (in fact, I think they're not yet part of C++ in general).

@mcy given you've been pushing them, I think you should open a PR against the style guide to allow them, and we can discuss it there. I think care has to be taken around overloading, in particular.